### PR TITLE
Stash integration (webhooks support and PR reporter)

### PR DESCRIPTION
**Pull request by hir12111**
# Main Features
- Support Bitbucket server webhook events (push, pull request created, pull request updated, pull request rejected and pull request fulfilled)
- Send customized (renderable) comments to a pull request in Bitbucket server

## Contributor Checklist:
* [x] I have updated the unit tests
* [x] I have created a file in the master/buildbot/newsfragment directory (and read the README.txt in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

**Comment by tardyp** [comment]
hi @hir12111, Thanks for your contribution. At first sight it looks pretty solid.

In order to facilitate a proper review, I would prefer that you submit two PRs, one for each component changed. Let me know if you need some help on the git fu to make that happen.

File: master/docs/manual/cfg-reporters.rst [hunk 1]
```diff
@@ -1012,6 +1012,48 @@ As a result, we recommend you use https in your base_url rather than http.
     :param boolean verify: disable ssl verification for the case you use temporary self signed certificates
     :param boolean debug: logs every requests and their response
 
+.. bb:reporter:: StashPRCommentPush
+
+StashPRCommentPush
```
Comment (hir12111) [inline master/docs/manual/cfg-reporters.rst]: I agree it would be a good idea to rename it as BitbucketServer, to distinguish it from the existing Bitbucket (cloud) APIs.

Do you have a preferred approach for keeping the old name available?

File: master/buildbot/www/hooks/bitbucketserver.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+# Copyright Mamba Team
+
+from __future__ import absolute_import
+from __future__ import print_function
+
+import json
+
+from twisted.python import log
+
+GIT_MERGE_REF = "refs/pull-requests/{}/merge"
```
Comment (hir12111) [inline master/buildbot/www/hooks/bitbucketserver.py]: yes, I've tested it in Bitbucket Server v4.14.4 and it works!

File: master/docs/manual/cfg-reporters.rst (hunk 1 above)
Comment (tardyp) [inline master/docs/manual/cfg-reporters.rst]:
```
class BitbucketServerStatusPush(..)
    [...]

def StashStatusPush(*args, **kwargs):
    warnings.warn("The 'StashStatusPush' class was renamed [...]",
                  DeprecationWarning )
    return BitbucketServerStatusPush(*args, **kwargs)
```

File: master/buildbot/reporters/stash.py [hunk 3]
```diff
@@ -61,8 +67,22 @@ def send(self, build):
...
             status = STASH_INPROGRESS
             description = self.startDescription
+
+        # got_revision could be a string, a dictionary or None
+        got_revision = props.getProperty('got_revision', None)
         for sourcestamp in build['buildset']['sourcestamps']:
             sha = sourcestamp['revision']
+
+            if sha is None:
+                if isinstance(got_revision, dict):
```
Comment (hir12111) [inline master/buildbot/reporters/stash.py]: In `handle_pullrequest` [line 130](https://github.com/buildbot/buildbot/pull/3314/files#diff-6a2f55ab8b3902c7f4433724cbce39ffR130) revision is explicitly set to `None` (meaning the latest revision), I thought this way is cleaner than calling Bitbucket server API or running git from the buildbot master server

Comment (tardyp) [inline master/buildbot/reporters/stash.py]: ah I did not see that. if the revision is available in the event we should set it.

If there is a new push, then we will get a new event, which in turn will trigger a new build testing that new sha.
The new event may cancel the previous build, this feature is not implemented, but for me, this is racy to expect the previous build to take latest version.

For all the github style pull request tester, (travis, appveyor) this is done like this, so I think we should stay coherent with this behaviour.

Comment (hir12111) [inline master/buildbot/reporters/stash.py]: I thought that before, unfortunately, the revision is not available in the payload on the pull request event, it is available the base branch and the compare branch revision but not the merge commit.
I saw that it is possible to get that commit revision by running git ls-remote {repository} refs/pull-requests/{PR}/merge, but this added many other problems (like needing authentication), what value would you suggest using for revision?

**Comment by bdbaddog** [comment]
Does this work with mercurial repo?
Or just git?

**Comment by hir12111** [comment]
Bitbucket Server doesn't support mercurial as (see [this](https://jira.atlassian.com/browse/BSERV-2469)), they state that the cost of supporting that would be too high

**Comment by bdbaddog** [comment]
Should this code also work with bitbucket.org? (which does support mercurial)

**Comment by tardyp** [comment]
@bdbaddog from what I understand, bitbucket.org and bitbucketserver are completly different product based on completly different codebase. Thus have different API.
This is a shame but we need to live with it..

**Comment by bdbaddog** [comment]
o.k. I'll press forward then to wrap up my changes for bitbucket webhook fixes.. ;)

File: master/docs/manual/cfg-reporters.rst (hunk 1 above)
Comment (hir12111) [inline master/docs/manual/cfg-reporters.rst]: done

File: master/buildbot/www/hooks/bitbucketserver.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        handler = getattr(self, 'handle_{}'.format(event_type), None)
+
+        if handler is None:
+            raise ValueError('Unknown event: {}'.format(event_type))
+
+        return handler(payload)
+
+    def _get_payload(self, request):
+        content = request.content.read()
+        content_type = request.getHeader(_HEADER_CT)
```
Comment (hir12111) [inline master/buildbot/www/hooks/bitbucketserver.py]: done

File: master/buildbot/www/hooks/bitbucketserver.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+    def handle_repo_push(self, payload):
+        changes = []
+        project = payload['repository']['project']['name']
+        repo_url = payload['repository']['links']['self'][0]['href']
+        repo_url = repo_url.rstrip('browse')
+
+        for payload_change in payload['push']['changes']:
+            if not payload_change['new']:
+                # skip change if deleting a tag or a branch
```
Comment (hir12111) [inline master/buildbot/www/hooks/bitbucketserver.py]: done, and I also changed the branch in case a tag is detected

File: master/buildbot/reporters/stash.py [hunk 6]
```diff
@@ -73,13 +93,57 @@ def send(self, build):
...
+    @defer.inlineCallbacks
+    def sendPullRequestComment(self, build):
+        props = Properties.fromDict(build['properties'])
+        pr_url = props.getProperty("pullrequesturl")
+        # we assume that the PR URL is well-formed as it comes from a PR event
+        path = urlparse(unicode2bytes(pr_url)).path
+        status = "SUCCESS" if build['results'] == SUCCESS else "FAILED"
+        props.setProperty('statustext', status, self.name)
+        props.setProperty('url', build['url'], self.name)
+        comment_text = yield props.render(self.text)
```
Comment (hir12111) [inline master/buildbot/reporters/stash.py]: I'm working on it, I suppose this may solve the warning issue

**Comment by hir12111** [comment]
Sorry about the status, I'm dealing with the problems that make it fail

**Comment by tardyp** [comment]
@hir12111 no problem. That is why we have a CI. its faster to just let it test the whole matrix :)

File: master/buildbot/newsfragments/reporters_stash.feature [hunk 7]
```diff
@@ -0,0 +1 @@
+Allow sending a comment to a pull request for Bibucket Server in :py:class:`~buildbot.reporters.stash.StashPRCommentPush`
```
Comment (tardyp) [inline master/buildbot/newsfragments/reporters_stash.feature]: BitbucketServerPTCommentPush

File: master/buildbot/www/hooks/bitbucketserver.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+from twisted.python import log
+
+GIT_BRANCH_REF = "refs/heads/{}"
+GIT_MERGE_REF = "refs/pull-requests/{}/merge"
+GIT_TAG_REF = "refs/tags/{}"
+
+_HEADER_EVENT = 'X-Event-Key'
+
+
+class BitbucketServerEventHandler(object):
```
Comment (tardyp) [inline master/buildbot/www/hooks/bitbucketserver.py]: Please have a look at https://github.com/buildbot/buildbot/pull/3327
It is a refactor to make the hooks more maintainable and flexible. you will need to rework that part when it lands

Comment (hir12111) [inline master/buildbot/www/hooks/bitbucketserver.py]: thank you for letting me know, I've taken a look at it, and it's not a problem to adapt it. I'll do it once that pr is merged.

Comment (tardyp) [inline master/buildbot/www/hooks/bitbucketserver.py]: it is merged now. Can you rebase?

File: master/buildbot/reporters/bitbucketserver.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        results = build['results']
+        if build['complete']:
+            status = SUCCESSFUL if results == SUCCESS else FAILED
+            description = self.endDescription
+        else:
+            status = INPROGRESS
+            description = self.startDescription
+
+        # got_revision could be a string, a dictionary or None
+        got_revision = props.getProperty('got_revision', None)
```
Comment (tardyp) [inline master/buildbot/reporters/bitbucketserver.py]: I think we shouldn't use got_revision in reporters. I removed that code from gitlab reporter for example. the change source should provide the revision.

File: master/buildbot/reporters/bitbucketserver.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+        for sourcestamp in build['buildset']['sourcestamps']:
+            sha = sourcestamp['revision']
+
+            if sha is None:
```
Comment (tardyp) [inline master/buildbot/reporters/bitbucketserver.py]: please look at gitlab reporter for a better boilerplate for reporters

File: master/buildbot/www/hooks/bitbucketserver.py (hunk 8 above)
Comment (hir12111) [inline master/buildbot/www/hooks/bitbucketserver.py]: done

File: master/buildbot/reporters/bitbucketserver.py (hunk 9 above)
Comment (hir12111) [inline master/buildbot/reporters/bitbucketserver.py]: maybe I should change the hook to use the source branch commit instead of None (for revision) on pull request events, the only problem I see is that the reporter will be updating the status of a different commit than the one being build (the merge commit)

File: master/buildbot/reporters/bitbucketserver.py (hunk 10 above)
Comment (hir12111) [inline master/buildbot/reporters/bitbucketserver.py]: if we assume that every chance source will set revision, can I just delete the check (if sha is None) ?

Comment (tardyp) [inline master/buildbot/reporters/bitbucketserver.py]: this could be a sourcestamp from another changesource, not bitbucket. So we should still support it and not send anything in that case

File: master/buildbot/reporters/bitbucketserver.py (hunk 9 above)
Comment (tardyp) [inline master/buildbot/reporters/bitbucketserver.py]: I think it is what we actually want. 
The CI checks for the source branch commits are:

- does that commit merges correctly with dest branch?
- does the resulting merged tree pass the unit tests?

Comment (hir12111) [inline master/buildbot/reporters/bitbucketserver.py]: done

File: master/buildbot/reporters/bitbucketserver.py (hunk 10 above)
Comment (hir12111) [inline master/buildbot/reporters/bitbucketserver.py]: I based on gitlab reporter, trying to be as much alike as possible

**Comment by tardyp** [comment]
I should have insisted on splitting it :)
thanks for the hard work!

**Comment by hir12111** [comment]
Thank you for the review, It was my pleasure to contribute to this awesome repository.
Next time, I'll make sure to split a big PR :)

**Comment by Beatris** [comment]
Hi,

The BitbucketServer API has been changed since the web hooks handler and the reporter have been implemented. Check the web hooks event payload (this is for v5 and there is also v6 already) - https://confluence.atlassian.com/bitbucketserver057/event-payload-945543761.html

Do you have a strategy to handle API changes in web hooks handlers?

We have implemented the new API with custom hook but it would be great if the web hooks have some versioning of the API they are handling.
